design/authenticators/authn_k8s/inject_client_cert_solution_design.md

```diff
@@ -0,0 +1,255 @@
+1. Calculate the SHA-1 hash of the cert content before sending it and then sending  a request to get the calculated SHA-1 hash of the cert in the container. Once the two hashes are equal then we know that the cert was fully installed.
+
+Note: we use `SHA-1` and not `MD5` as the latter is not FIPS compliant. 
+
+#### Option 2 - Send the cert content as part of the URL
+In this option, we do not use STDIN to send the cert but send it in the URL. This eliminates the bug in KubeCtl Exec as without STDIN we will get the response and would be able to know whether the injection succeeded or not, and log the outcome.
+
+However, we may have a limit on the size of the URL and we will have an issue  if the certificate will grow in the future. To overcome this, we can split the  request. The first one will create the file. The following requests will populate it with the cert content. Then the last request will verify that the whole content has passed (with one of the methods described above). We can decide on the number  of requests dynamically according to the size of the cert.
+
+We have a security concern with this option. While the traffic is fully encrypted when using https, including the URL, [it's still possible that the URL will be visible in certain locations](https://owasp.org/www-community/vulnerabilities/Information_exposure_through_query_strings_in_url), while the request travels to its destination. It is possible that the Kubernetes API server will not leak any information from the URL, but if the Conjur instance runs outside the cluster, it will surely pass some proxies and load balancers before it reaches Kubernetes.
+
+Thus, we will not use this option due to security concerns.
+
+#### Option 3 - Client sets its State once cert is injected
+This option is relevant only if the "inject-client-cert" request will return a 202 Accepted response instead of 200 OK (more on that in [this section](#inject-client-cert-response-code).
+
+In this case, we respond with 202 Accepted to the "inject-client-cert" request and start injecting the cert in a separate thread. The server then polls on the state of the authn-client container and waits until its State changes to some  agreed State (e.g "CertInjected"). This state will be set by the authn-client once it sees that the cert is installed.
+
+This option is not fully backwards-compatible and is quite complex so I wouldn't suggest to implement it.
+
+#### Option 4 - Send cert as part of bash script
+
+The issue with the STDIN that is not closing is an issue of the `tar` command that waits for more input and doesn't close the connection. Another option that we have is to change the command from `tar` to `sh` and send a bash script in the STDIN.
+The script can be something like:
+
+```
+#!/bin/sh
+
+cat > /etc/conjur/ssl/client.pem <<EOF
+#{cert-content}
+EOF
+
+chmod 644 /etc/conjur/ssl/client.pem
+exit
+```
```

Comment:
should this be updated to what you shared today?
```
#!/bin/sh
copy_file() {
  cat > #{path} <<EOF
#{content}
EOF
  chmod #{mode} #{path}
}
copy_file 2>&1 | tee -a #{COPY_FILE_LOG_FILE}
exit
```

also, this looks to be connected to the open PR #1802 - I'd like to make sure we factor in that the cert may be injected at a configurable path in the near future (cc @JakeQuilty)

Reply:
yes, thanks. i updated the script.
I saw @JakeQuilty 's work and it looks like they can both live together. there will be some merge conflicts so whoever finishes first can help the other one with the merge. @JakeQuilty - let's be in contact once one of us merges their work so we can perform the merge together if needed.